scripts/restore_plugins_snapshot

```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+##--------------------------------------------------------------------
+## Copyright (c) 2017-2018 OSIsoft, LLC
+##
+## Licensed under the Apache License, Version 2.0 (the "License");
+## you may not use this file except in compliance with the License.
+## You may obtain a copy of the License at
+##
+##     http://www.apache.org/licenses/LICENSE-2.0
+##
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS,
+## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+## See the License for the specific language governing permissions and
+## limitations under the License.
+##--------------------------------------------------------------------
+
+set -e
+#set -x
+
+#
+# This is the plugins snapshot restore scripts
+
+# File Name
+if [ -z ${1+x} ]; then
+  exit 1
+else
+  FILE_NAME="$1"
+fi
+
+sudo /bin/tar -C / -xzf "$FILE_NAME"
```

Comment:
If this is a setuid on execution script we do not need the sudo in this line.

Reply:
Regret to inform that this setuid thing is not working. The earlier "success" was due to the fact that the script restore_snapshot_plugins contained a sudo /bin/tar command and before executing this script, I had probably run a sudo command due to which the sudo password remained active for some time. I deeply regret the slip.

Secondly, the setuid for a script is considered a security risk and there are plenty of reference material available on the net advising against it, A few such reference urls are:
https://unix.stackexchange.com/questions/364/allow-setuid-on-shell-scripts
http://www.faqs.org/faqs/unix-faq/faq/part4/section-7.html

Thereafter, I decided to test the setuid bit on a test script as below. I created a folder test, owned by root, and created a shell script, also owned by root but having its setuid bit set. When executed by a normal user, asinha, it did not work and threw permission error. 
```
~ $ whoami
asinha
~ $ 
~ $ ll -d test
drwxr-xr-x 2 root root 4096 Apr 11 21:07 test/
~ $ 
~ $ cd test
test $ 
test $ ll
total 12
drwxr-xr-x  2 root   root   4096 Apr 11 21:07 ./
drwxr-xr-x 32 asinha asinha 4096 Apr 11 21:06 ../
-rwSr-Sr-x  1 root   root     34 Apr 11 21:00 scrpt*
test $ 
test $ cat scrpt 
#!/bin/bash
mkdir test2
cd test2

test $ 
test $ whoami
asinha
test $ 
test $ ./scrpt 
mkdir: cannot create directory ‘test2’: Permission denied
./scrpt: line 3: cd: test2: No such file or directory
test $ 
test $
```
Therefore, it seems that we have no other option but to revert back to entries in foglamp.sudoers file. Please advise.

Reply:
We can probably write a small C helper program to do tar archive extract in '/' and set setuid bit on it.

Reply:
@AmandeepArora to check this, can we create a quick C executable to replicate the scrpt script file above?

Reply:
@AmandeepArora @MarkRiddoch a C++ executable, test, provided by @AmandeepArora seems to be working fine and thus serves our purpose as below:
```
FogLAMP (FOGL-2655 %) $ ll -d test
drwxr-xr-x 2 root root 4096 Apr 15 12:03 test/
FogLAMP (FOGL-2655 %) $ 
FogLAMP (FOGL-2655 %) $ 
FogLAMP (FOGL-2655 %) $ cd test
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ ll
total 20
drwxr-xr-x  2 root   root   4096 Apr 15 12:03 ./
drwxrwxr-x 23 asinha asinha 4096 Apr 15 11:55 ../
-rwsr-sr-x  1 root   root   8824 Apr 15 11:55 test*
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ whoami
asinha
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ ./test
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ ll
total 24
drwxr-xr-x  3 root   root   4096 Apr 15 12:04 ./
drwxrwxr-x 23 asinha asinha 4096 Apr 15 11:55 ../
-rwsr-sr-x  1 root   root   8824 Apr 15 11:55 test*
drwxrwx---  2 root   root   4096 Apr 15 12:04 test2/
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ 
test (FOGL-2655 %) $ whoami
asinha
test (FOGL-2655 %) $ 
```

Hence it is proposed to go ahead with the C helper program to extract a .tar.gz file as proposed by @AmandeepArora above. This C++ helper program will receive .tar.gz filename as a command line argument.

Thanks @AmandeepArora for this initiative.

Reply:
I provided this C binary for tar extraction. There is still a major security risk involved. The program would extract the compressed archive by setting "-C /", so in a way, it could potentially be used to even overwrite binaries in /bin, /usr/bin/, /sbin, etc. May be there should be further verification of tar archive contents in this C binary program before extraction.

Reply:
Good, let's go with a small utility program to do it. Seems better than having to use sudo to me. Seems much more secure.